8263322: Calling Application.launch on FX thread should throw IllegalStateException, but causes deadlock

FlorianKirmaier opened the pull request:
Fixing deadlock when calling Application.launch in the FXThread after Platform.startup

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263322](https://bugs.openjdk.java.net/browse/JDK-8263322): Calling Application.launch on FX thread should throw IllegalStateException, but causes deadlock


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/421/head:pull/421` \
`$ git checkout pull/421`

Update a local copy of the PR: \
`$ git checkout pull/421` \
`$ git pull https://git.openjdk.java.net/jfx pull/421/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 421`

View PR using the GUI difftool: \
`$ git pr show -t 421`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/421.diff">https://git.openjdk.java.net/jfx/pull/421.diff</a>

</details>

kevinrushforth commented:
/reviewers 2

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/javafx/application/LauncherImpl.java:
```diff
@@ -171,7 +171,7 @@ public static void launchApplication(final Class<? extends Application> appClass
             final String[] args) {
 
-        if (launchCalled.getAndSet(true)) {
+        if (com.sun.glass.ui.Application.isEventThread() || launchCalled.getAndSet(true)) {
```
Can you do this as two separate `if` checks, with the `isEventThread` check first? The error message for this case should be something like "Application launch must not be called on the JavaFX Application Thread".

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+package test.javafx.scene;
```
1. Missing Copyright header block
2. Other platform startup tests are in `test.com.sun.javafx.application`; can you move this test there as well?

FlorianKirmaier replied:
done

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    public static void initializeApplication() throws Exception {
```
This method is unused, along with the `InitializeApp` class. Did you plan to use it?

FlorianKirmaier replied:
It's useful to compare it with the behavior of the two methods to start JavaFX.
If it would be my codebase, I would keep it, so if someone investigates it later, it's easier to investigate for differences. But I can also delete it, what would you prefer?

FlorianKirmaier replied:
It's now used because i added more test cases.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+        new Thread(() -> {
+            Application.launch(InitializeApp.class);
+        }).start();
+        appLatch.await();
```
Presuming you intend to use this (as opposed to removing it), this should be changed to an await with a timeout.

FlorianKirmaier replied:
done

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+        Platform.startup(() -> {
+            latch.countDown();
+        });
+        latch.await();
```
This needs to be changed to a flavor of await with a timeout (you can assert that it doesn't timeout). Also, I don't think this needs to be its own method, since the only thing the `initialize` method does is call this.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+                latch.countDown();
+            } catch (Exception e) {
+                System.out.println("got exception:  " + e);
+                e.printStackTrace();
```
You should rethrow the exception in this case, since it indicates an unexpected error. Better still, you can remove this block and not catch the Exception in the first place.

FlorianKirmaier replied:
Not relevant anymore, because of Util.runAndWait()

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    @Test
```
All of the test methods should take a timeout parameter: `@Test (timeout = 15000)`

FlorianKirmaier replied:
done

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+            System.out.println("Calling launch!");
+            Application.launch(TestApp.class);
+            System.out.println("Finished launch!");
+            throw new Exception("We excpect an error!");
```
Usual practice in this case would be to use the junit `Assert.fail` method (or else `throw new AssertionError(...)`).

FlorianKirmaier replied:
done

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+    @Test
+    public void testStartupThenLaunchInFX() throws Exception {
+        CountDownLatch latch = new CountDownLatch(1);
+        Platform.runLater(() -> {
```
I recommend using `Util.runAndWait` since it will propagate exceptions from the `runLater` lambda to the caller.

FlorianKirmaier replied:
This significantly simplifies the test!

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/InitializeJavaFXTest.java:
```diff
@@ -0,0 +1,84 @@
+                e.printStackTrace();
+            }
+        });
+        Assert.assertTrue("Timeout", latch.await(10, TimeUnit.SECONDS));
```
If you use `runAndWait` you won't need this.

FlorianKirmaier replied:
Done

kevinrushforth reviewed (changes requested):
The idea behind the fix is good. A few changes are needed:

1. The check should be split into two separate `if` statements, each with their own error message (see inline).

2. This should be documented in the API docs for the Application::launch method as an additional case that will throw an IllegalStateException. It will need a reasonably trivial CSR since we are specifying a new case that will cause an exception to be thrown.

3. The system test need to be broken up into separate files, one for each `@Test` method, since application launching tests need to run in their own JVM. If you want to share any code, you could refactor it into a common class that has methods (not annotated with `@Test`) to perform the testing, and separate classes that will subclass the common class, each with a single `@Test` method that simply calls into the method in the common class to do the testing. See any number of examples in `tests/system/src/test/java/test/com/sun/javafx/application/` (which is also a better location for your new test). You will want to add a cleanup method annotated with `@AfterClass` that calls `Platform.exit()`. I think three tests would be good:
    * Initalize the FX runtime via Platform.startup and then launch an Application on another thread (should succeed)
    * Initalize the FX runtime via Platform.startup and then launch an Application on another thread (should throw ISE)
    * Initalize the FX runtime via Application.launch and then launch a second Application on another thread (should throw ISE).


I note that when I run your test, it hangs (on Windows...I didn't try it on other platforms).

```
gradle --info -PFULL_TEST=true cleanTest :systemTests:test --test InitializeJavaFXTest
```

This might be resolved by ensuring each test is in its own JVM as requested above.


Additional comments are inline.

kevinrushforth commented:
/csr

kevinrushforth commented:
> * Initalize the FX runtime via Platform.startup and then launch an Application on another thread (should throw ISE)

That should be:

* Initalize the FX runtime via Platform.startup and then launch an Application on the FX Application thread (should throw ISE)

FlorianKirmaier commented:
Contribution.MD states, that it's automatically checked for whitespace errors? Is this not true?

As a clarification, this PR restores the previous behavior before Platform.startup. With this PR Application.launch and Platform.startup behaves the same.

FlorianKirmaier commented:
I forgot to commit a part of the fix, which is why the second test hang. It's now part of the PR.

FlorianKirmaier commented:
About the CSR:
This commit actually restores original behavior, when JavaFX was called with Application.launch, so it's not really a change in the API, it's more like a fix for a regression.
How would I create a CSR? Just by creating a ticket at https://bugs.openjdk.java.net/issues/?jql=issuetype%20%3D%20CSR with the type CSR?

FlorianKirmaier commented:
`Initalize the FX runtime via Platform.startup and then launch an Application on another thread (should succeed)`
I don't think that should succeed. I would expect it to throw an exception. If that would be the case, then both startup-methods would result in different states, which wouldn't be so good.

I've now restructured the tests.
It's a total of 4 tests. 2 after Application.launch and 2 after Platform.startup. 
In both cases I check what happens, a Application is launched from another thread, or from the javafx thread. In all cases an exception is expected.

kevinrushforth commented:
> Contribution.MD states, that it's automatically checked for whitespace errors? Is this not true?

Yes, Skara's jcheck does basic sanity checking (tabs, trailing white space, line endings). Why do you ask? I didn't raise this as a question in this PR. Do you have reason to believe that this isn't working?

> As a clarification, this PR restores the previous behavior before Platform.startup. With this PR Application.launch and Platform.startup behaves the same.

Not quite. `Platform.startup` is a new method that was previously available only via an internal implementation class, which simply starts up the FX runtime without running any of the application life cycle. Once this was added to public API, it became possible (but is seldom necessary) to call `Platform.startup` and then later call `Application.launch`, which should work fine as long as the latter call is _not_ on the FX Application Thread. What was missed is a check to make sure that `Application.launch` is not called on the FX Application thread. This should have been documented to throw an exception. It is somewhat similar to, but not the same as, the case of calling `Application.launch` more than once.

kevinrushforth commented:
> Initalize the FX runtime via Platform.startup and then launch an Application on another thread (should succeed)
I don't think that should succeed. I would expect it to throw an exception. If that would be the case, then both startup-methods would result in different states, which wouldn't be so good.

No, this really should succeed. Internally, it is a similar case to what the special Java launcher method does when launching a Java class that extends `Application` and which may have a main method that calls `Application.launch`. We check the various cases of launching an application with/without it extending `Application` in the tests under [tests/system/src/test/java/test/launchertest/](https://github.com/openjdk/jfx/tree/master/tests/system/src/test/java/test/launchertest/).

kevinrushforth commented:
> About the CSR:
> This commit actually restores original behavior, when JavaFX was called with Application.launch, so it's not really a change in the API, it's more like a fix for a regression.

Not quite. See  my reply above.

> How would I create a CSR? Just by creating a ticket at https://bugs.openjdk.java.net/issues/?jql=issuetype%20%3D%20CSR with the type CSR?

Almost. Go to your bug in JBS, and use the "More" pull-down menu. There is a "Create CSR" option.

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/javafx/application/LauncherImpl.java:
```diff
@@ -658,7 +658,7 @@ private static void startToolkit() throws InterruptedException {
 
         // Note, this method is called on the FX Application Thread
-        PlatformImpl.startup(() -> startupLatch.countDown(), true);
+        PlatformImpl.startup(() -> startupLatch.countDown());
```
Glad to see this reverted. I was going to ask you why it was needed (it shouldn't be).

FlorianKirmaier replied:
It was added, because I thought it shouldn't be allowed to call Application.launch after Platform.startup.

kevinrushforth commented:
Regarding the CSR, it is only concerned with the public API, including the specification (javadoc-generated API documentation), and any behavioral changes.

You can leave the CSR in the Draft state until there is something to document in the Specification section, and the review of the PR is far enough along so that those changes are likely to be final. For this PR, the changes that need to be documented in the CSR haven't been made yet. The changes needed will be in the javadoc comments of public `Application::launch` methods.

FlorianKirmaier commented:
I've now added the new Exception to the JavaDoc of the Application.

kevinrushforth commented:
The API spec changes look good. You can add that to the Specification section of the CSR. You can either paste the diffs for the javadoc comments, or else just add the one new line for each method. Either way, the name of the class (javafx.application.Application) should be listed prior to the diffs / additions. Also, make it clear which methods are getting the new `@exception` tags.

FlorianKirmaier commented:
Great, I've updated the CSR!

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXStartupTest.java:
```diff
@@ -0,0 +1,53 @@
+        System.out.println("Called Startup!");
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunchInFX() throws Exception {
+        doTestInitializeThenLaunchInFX();
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunch() throws Exception {
```
It will be more robust for these two test methods to be in separate files. Otherwise they might interfere with each other. If `testStartupThenLaunchInFX` runs first, then an error in that test (e.g., if the fix isn't applied and it times out) will cause the `testStartupThenLaunch` to fail. Conversely, if `testStartupThenLaunch` runs first, then the system is in a state where the Application has already been started by the time `testStartupThenLaunchInFX` runs. It will still pass, but won't be a solid test of the fix, since that case would fail today.

FlorianKirmaier replied:
They are seperated now!

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXLaunchTest.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.println("Called Startup!");
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunchInFX() throws Exception {
+        doTestInitializeThenLaunchInFX();
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunch() throws Exception {
```
It will be more robust for these two test methods to be in separate files. Otherwise they might interfere with each other.

FlorianKirmaier replied:
They are seperated now!

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXBase.java:
```diff
@@ -0,0 +1,98 @@
+                Application.launch(TestApp.class);
+                System.out.println("Finished launch!");
+                Assert.fail("Error: No Exception was thrown - expected IllegalStateException");
+            } catch (IllegalStateException e) {
```
Maybe add a comment that this exception is expected?

FlorianKirmaier replied:
I've added a comment and a print-statement that this is expected - on both catches for EllegalStateException

kevinrushforth replied:
The comment looks fine. See my global comment about the print statement.

kevinrushforth reviewed (commented):
The fix looks good. I have a couple comments on the tests, mainly about making them more robust by having only a single `@Test` method in each file.

arapte commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXLaunchTest.java:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2012, 2021, Oracle and/or its affiliates. All rights reserved.
```
Minor year correction-> `Copyright (c) 2021,`

FlorianKirmaier replied:
Done, so it's the year where changes happened, ok!

kevinrushforth replied:
In general, the pattern is: `year_created, year_last_modified,` (with the last modified year omitted if it's the same as the creation year).

arapte commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXBase.java:
```diff
@@ -0,0 +1,98 @@
+import javafx.application.Platform;
+import javafx.stage.Stage;
+import junit.framework.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
```
unused imports, BeforeClass and Test.

FlorianKirmaier replied:
removed!

arapte reviewed (commented):
Suggested minor changes.

FlorianKirmaier commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXBase.java:
```diff
@@ -0,0 +1,98 @@
+            System.out.println("Works!");
+        } catch (Exception e) {
+            throw e;
+        }
```
Line 91 is reachable. Application.launch throws an IllegalStateException, even so, it doesn't declare it.
I've Removed the second catch block.

kevinrushforth replied:
At the time Ambarish made the comment, line 91 referred to this statement:

```
             throw new Exception();
```

This is indeed not reachable (although the compiler can't tell that), since `Assert.fail` unconditionally throws an `AssertionError`

FlorianKirmaier commented:
I've seperated the tests now in seperated files!
Now all the changes requested should be in the PR.

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXBase.java:
```diff
@@ -0,0 +1,69 @@
+
+public class InitializeJavaFXBase {
+
+
```
Minor: there is an extra blank line here

FlorianKirmaier replied:
removed

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXLaunchBase.java:
```diff
@@ -0,0 +1,24 @@
+        new Thread(() -> {
+            Application.launch(InitializeApp.class);
+        }).start();
+        appLatch.await(5, TimeUnit.SECONDS);
```
You should `assertTrue` the return value of `await`.

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXStartupBase.java:
```diff
@@ -0,0 +1,17 @@
+        Platform.startup(() -> {
+            latch.countDown();
+        });
+        latch.await(5, TimeUnit.SECONDS);
```
You should `assertTrue` the return value of `await`.

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXLaunch1Test.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunchInFX() throws Exception {
```
Would something like `testLaunchThenLaunchInFX` be a better name? As it is, it uses the same name as the test that really does use `Platform.startup` which I found confusing.

FlorianKirmaier replied:
Yes, I've fixed the method names!

kevinrushforth commented on tests/system/src/test/java/test/com/sun/javafx/application/InitializeJavaFXLaunch2Test.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Test (timeout = 15000)
+    public void testStartupThenLaunch() throws Exception {
```
Would something like `testLaunchThenLaunch` be a better name?

kevinrushforth reviewed (commented):
Thanks for splitting up the tests. I ran them with the fix and all pass, I then ran them without the fix and only the one I expected to fail did.

Before approving this, can you change the title of the JBS bug and CSR issue to match (after doing that you will also need to update this PR title) as recommended in the CSR? My suggestion is something like:

```
Calling Application.launch on FX thread should throw IllegalStateException, but causes deadlock
```

Another thing I recommend is to comment out or remove the print statements in the tests. They are useful while debugging, but in production we prefer tests to only print something as a warning or error. 

I also left a few inline comments.

FlorianKirmaier commented:
I've made all the changes in the comments, and also changed the names of the tickets and the PR.
For some reason, I can't interact with the two unclosed comments.
I've removed the unreachable code, and the print-statements are also removed.

kevinrushforth reviewed (approved):
Looks good.

The CSR is ready to move to the Proposed state (there is one typo in the description that I pointed out).

FlorianKirmaier commented:
The CSR is now in the proposed state!

kevinrushforth commented:
I reviewed the CSR, and it is now ready for you to Finalize.

kevinrushforth commented:
@FlorianKirmaier The CSR needs to be updated (fixVersion set to `openjfx17`) and moved to Finalize again.

FlorianKirmaier commented:
The fixVersion is now openfjx17 and it's moved to "Finalized"

FlorianKirmaier commented:
/integrate

kevinrushforth commented:
/sponsor

kevinrushforth commented:
Looks like the Skara bot is in a loop continually reprocessing the `sponsor` command it already processed. I filed [SKARA-956](https://bugs.openjdk.java.net/browse/SKARA-956) to track this.
